state/address.go

```diff
@@ -97,39 +102,92 @@ type apiHostPortsDoc struct {
+		var existingAllAddrDoc apiHostPortsDoc
+		err := controllers.Find(bson.D{{"_id", apiHostPortsKey}}).One(&existingAllAddrDoc)
 		if err != nil {
 			return nil, err
 		}
-		op := txn.Op{
-			C:  controllersC,
-			Id: apiHostPortsKey,
-			Assert: bson.D{{
-				"txn-revno", existingDoc.TxnRevno,
-			}},
+		hostPorts := networkHostsPorts(existingAllAddrDoc.APIHostPorts)
+
+		// If the stored addresses are different from the input collection, queue an update operation.
+		if !hostsPortsEqual(newHostPorts, hostPorts) {
+			ops = append(ops, txn.Op{
+				C:  controllersC,
+				Id: apiHostPortsKey,
+				Assert: bson.D{{
+					"txn-revno", existingAllAddrDoc.TxnRevno,
+				}},
+				Update: bson.D{{
+					"$set", bson.D{{"apihostports", fromNetworkHostsPorts(newHostPorts)}},
+				}},
+			})
+		}
+
+		// Filter the collection of API addresses based on the configured management
+		// space for the controller.
+		// If there is no space configured, or if the one of the slices is filtered down
+		// to zero elements, just use the unfiltered slice for safety.
+		config, err := st.ControllerConfig()
+		if err != nil {
+			return nil, err
 		}
-		hostPorts := networkHostsPorts(existingDoc.APIHostPorts)
-		if !hostsPortsEqual(netHostsPorts, hostPorts) {
-			op.Update = bson.D{{
-				"$set", bson.D{{"apihostports", doc.APIHostPorts}},
-			}}
+		if mgmtSpace := config.JujuManagementSpace(); mgmtSpace != "" {
+			newHostPortsForAgents = make([][]network.HostPort, len(newHostPorts))
+			sp := network.SpaceName(mgmtSpace)
+			for i := range newHostPorts {
+				if filtered, ok := network.SelectHostsPortBySpaces(newHostPorts[i], sp); ok {
+					newHostPortsForAgents[i] = filtered
+				} else {
+					newHostPortsForAgents[i] = newHostPorts[i]
+				}
+			}
 		} else {
+			newHostPortsForAgents = newHostPorts
+		}
+
+		// Retrieve the current API addresses for agents document.
+		var existingAgentAddrDoc apiHostPortsDoc
+		err = controllers.Find(bson.D{{"_id", apiHostPortsForAgentsKey}}).One(&existingAgentAddrDoc)
+		if err != nil {
+			return nil, err
+		}
+
+		// If the stored addresses are different from the new management API address
+		// collection, queue an update operation.
+		agentHostPorts := networkHostsPorts(existingAgentAddrDoc.APIHostPorts)
+		if !hostsPortsEqual(newHostPortsForAgents, agentHostPorts) {
+			ops = append(ops, txn.Op{
+				C:  controllersC,
+				Id: apiHostPortsForAgentsKey,
+				Assert: bson.D{{
+					"txn-revno", existingAgentAddrDoc.TxnRevno,
+				}},
+				Update: bson.D{{
+					"$set", bson.D{{"apihostports", fromNetworkHostsPorts(newHostPortsForAgents)}},
+				}},
+			})
+		}
+
+		if ops == nil || len(ops) == 0 {
 			return nil, statetxn.ErrNoOperations
 		}
-		return []txn.Op{op}, nil
+		return ops, nil
 	}
 	if err := st.db().Run(buildTxn); err != nil {
 		return errors.Annotate(err, "cannot set API addresses")
 	}
-	logger.Debugf("setting API hostPorts: %v", netHostsPorts)
+	logger.Debugf("setting API hostPorts: %v", newHostPorts)
```

Comment:
Can you play around with this a bit and see if it is understandable when you actually see it in the log file?
Ideally we wouldn't log 2x, but we also want to be careful about the type structures, and ensure that we're really printing debug information that is understandable to someone who is trying to read them.
One option is also to try out:
logger.Debugf("... %# v", pretty.Formatter(newHostPortsForAgents))

using
github.com/kr/pretty
which would give pretty-printed lists (with line wrapping, indentation, etc.)

We also don't really want to spam the log files, so some care needs to be between verbosity and ease of reading.